lib/types.nix

```diff
@@ -246,7 +246,25 @@ rec {
     either = t1: t2: mkOptionType {
       name = "${t1.name} or ${t2.name}";
       check = x: t1.check x || t2.check x;
-      merge = mergeOneOption;
+      merge = loc: defs:
+             if all t1.check (getValues defs) then t1.merge loc defs
```

Comment:
I will recommend to go for  `all t1.check values && all (x: ! t2.check x) values`  in the condition, and identically for the second condition.